ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -2226,30 +2229,314 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
+		return PKCS11_CKR_GENERAL_ERROR;
+	}
+
+	mbedtls_pk_init(&pk);
+	mbedtls_mpi_init(&n);
+	mbedtls_mpi_init(&e);
+	mbedtls_mpi_init(&d);
+	mbedtls_mpi_init(&p);
+	mbedtls_mpi_init(&q);
+
+	mbedtls_rc = mbedtls_pk_parse_key(&pk, data, key_size, NULL, 0);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rsa = mbedtls_pk_rsa(pk);
+	mbedtls_rc = mbedtls_rsa_export(rsa, &n, &p, &q, &d, &e);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (key_bits && mbedtls_mpi_bitlen(&n) != key_bits) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
+		goto out;
+	}
+
+	size = ROUNDUP(mbedtls_mpi_bitlen(&n), 8) / 8;
+	buffer_size = size;
+	buffer = TEE_Malloc(buffer_size, TEE_USER_MEM_HINT_NO_FILL_ZERO);
+	if (!buffer) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_mpi_write_binary(&n, buffer, size);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rc = add_attribute(head, PKCS11_CKA_MODULUS, buffer, size);
+	if (rc)
+		goto out;
+
+	size = ROUNDUP(mbedtls_mpi_bitlen(&e), 8) / 8;
+	if (buffer_size < size) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_mpi_write_binary(&e, buffer, size);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rc = add_attribute(head, PKCS11_CKA_PUBLIC_EXPONENT, buffer, size);
+	if (rc)
+		goto out;
+
+	size = ROUNDUP(mbedtls_mpi_bitlen(&d), 8) / 8;
+	if (buffer_size < size) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_mpi_write_binary(&d, buffer, size);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rc = add_attribute(head, PKCS11_CKA_PRIVATE_EXPONENT, buffer, size);
+	if (rc)
+		goto out;
+
+	size = ROUNDUP(mbedtls_mpi_bitlen(&p), 8) / 8;
+	if (buffer_size < size) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_mpi_write_binary(&p, buffer, size);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rc = add_attribute(head, PKCS11_CKA_PRIME_1, buffer, size);
+	if (rc)
+		goto out;
+
+	size = ROUNDUP(mbedtls_mpi_bitlen(&q), 8) / 8;
+	if (buffer_size < size) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_mpi_write_binary(&q, buffer, size);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rc = add_attribute(head, PKCS11_CKA_PRIME_2, buffer, size);
+
+out:
+	mbedtls_pk_free(&pk);
+	mbedtls_mpi_free(&n);
+	mbedtls_mpi_free(&e);
+	mbedtls_mpi_free(&d);
+	mbedtls_mpi_free(&p);
+	mbedtls_mpi_free(&q);
+	TEE_Free(buffer);
+	return rc;
+}
+
 enum pkcs11_rc set_key_data(struct obj_attrs **head, void *data,
 			    size_t key_size)
 {
+	enum pkcs11_key_type type = 0;
+
 	switch (get_class(*head)) {
 	case PKCS11_CKO_SECRET_KEY:
 		return set_secret_key_data(head, data, key_size);
+	case PKCS11_CKO_PRIVATE_KEY:
+		type = get_key_type(*head);
+		if (type == PKCS11_CKK_RSA)
```

Comment:
can remove `type` and use `if (get_key_type(head) == PKCS11_CKK_RSA)` straight.

Reply:
fixed